Deamon rhsmcertd is able to install certs for docker again

jirihnidek opened the pull request:
* BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1671534
* Deamon rhsmcertd wasn't able to install entitlement certificates
  for docker. More details are in bug report. This change fixes
  this issue.

jirihnidek committed:
Deamon rhsmcertd is able to install certs for docker again

* BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1671534
* Deamon rhsmcertd wasn't able to install entitlement certificates
  for docker. More details are in bug report. This change should
  fix this issue.

wrabcak commented:
Hi, 
Could you please provide SELinux denial (AVC msg) ? 

Thanks,
Lukas.

jirihnidek commented:
Hi @wrabcak ,
here are AVC denials:
```
----
time->Thu Apr 11 10:29:57 2019
type=AVC msg=audit(1554971397.059:322): avc:  denied  { search } for  pid=2661 comm="rhsmcertd-worke" name=".local" dev="dm-0" ino=9044513 scontext=system_u:system_r:rhsmcertd_t:s0 tcontext=unconfined_u:object_r:gconf_home_t:s0 tclass=dir permissive=0
----
time->Thu Apr 11 10:29:57 2019
type=AVC msg=audit(1554971397.481:323): avc:  denied  { search } for  pid=2661 comm="rhsmcertd-worke" name="docker" dev="dm-0" ino=11345213 scontext=system_u:system_r:rhsmcertd_t:s0 tcontext=system_u:object_r:container_config_t:s0 tclass=dir permissive=0
----
time->Thu Apr 11 10:29:57 2019
type=AVC msg=audit(1554971397.482:324): avc:  denied  { search } for  pid=2661 comm="rhsmcertd-worke" name="docker" dev="dm-0" ino=11345213 scontext=system_u:system_r:rhsmcertd_t:s0 tcontext=system_u:object_r:container_config_t:s0 tclass=dir permissive=0
----
time->Thu Apr 11 10:29:57 2019
type=AVC msg=audit(1554971397.482:325): avc:  denied  { search } for  pid=2661 comm="rhsmcertd-worke" name="docker" dev="dm-0" ino=11345213 scontext=system_u:system_r:rhsmcertd_t:s0 tcontext=system_u:object_r:container_config_t:s0 tclass=dir permissive=0
----
time->Thu Apr 11 10:29:57 2019
type=AVC msg=audit(1554971397.483:326): avc:  denied  { search } for  pid=2661 comm="rhsmcertd-worke" name="docker" dev="dm-0" ino=11345213 scontext=system_u:system_r:rhsmcertd_t:s0 tcontext=system_u:object_r:container_config_t:s0 tclass=dir permissive=0
----
time->Thu Apr 11 10:30:00 2019
type=AVC msg=audit(1554971400.494:327): avc:  denied  { noatsecure } for  pid=2667 comm="dbus-daemon-lau" scontext=system_u:system_r:system_dbusd_t:s0-s0:c0.c1023 tcontext=system_u:system_r:setroubleshootd_t:s0-s0:c0.c1023 tclass=process permissive=0
----
time->Thu Apr 11 10:30:00 2019
type=AVC msg=audit(1554971400.495:328): avc:  denied  { rlimitinh } for  pid=2667 comm="setroubleshootd" scontext=system_u:system_r:system_dbusd_t:s0-s0:c0.c1023 tcontext=system_u:system_r:setroubleshootd_t:s0-s0:c0.c1023 tclass=process permissive=0
----
time->Thu Apr 11 10:30:00 2019
type=AVC msg=audit(1554971400.495:329): avc:  denied  { siginh } for  pid=2667 comm="setroubleshootd" scontext=system_u:system_r:system_dbusd_t:s0-s0:c0.c1023 tcontext=system_u:system_r:setroubleshootd_t:s0-s0:c0.c1023 tclass=process permissive=0
----
time->Thu Apr 11 10:30:07 2019
type=AVC msg=audit(1554971407.625:330): avc:  denied  { noatsecure } for  pid=2697 comm="virt-what" scontext=system_u:system_r:rhsmcertd_t:s0 tcontext=system_u:system_r:dmidecode_t:s0 tclass=process permissive=0
----
time->Thu Apr 11 10:30:07 2019
type=AVC msg=audit(1554971407.625:331): avc:  denied  { rlimitinh } for  pid=2697 comm="dmidecode" scontext=system_u:system_r:rhsmcertd_t:s0 tcontext=system_u:system_r:dmidecode_t:s0 tclass=process permissive=0
----
time->Thu Apr 11 10:30:07 2019
type=AVC msg=audit(1554971407.625:332): avc:  denied  { siginh } for  pid=2697 comm="dmidecode" scontext=system_u:system_r:rhsmcertd_t:s0 tcontext=system_u:system_r:dmidecode_t:s0 tclass=process permissive=0
----
time->Thu Apr 11 10:30:07 2019
type=AVC msg=audit(1554971407.672:333): avc:  denied  { noatsecure } for  pid=2734 comm="virt-what" scontext=system_u:system_r:rhsmcertd_t:s0 tcontext=system_u:system_r:dmidecode_t:s0 tclass=process permissive=0
----
time->Thu Apr 11 10:30:07 2019
type=AVC msg=audit(1554971407.672:334): avc:  denied  { rlimitinh } for  pid=2734 comm="dmidecode" scontext=system_u:system_r:rhsmcertd_t:s0 tcontext=system_u:system_r:dmidecode_t:s0 tclass=process permissive=0
----
time->Thu Apr 11 10:30:07 2019
type=AVC msg=audit(1554971407.672:335): avc:  denied  { siginh } for  pid=2734 comm="dmidecode" scontext=system_u:system_r:rhsmcertd_t:s0 tcontext=system_u:system_r:dmidecode_t:s0 tclass=process permissive=0
----
time->Thu Apr 11 10:30:07 2019
type=AVC msg=audit(1554971407.982:336): avc:  denied  { search } for  pid=2764 comm="rhsmcertd-worke" name=".local" dev="dm-0" ino=9044513 scontext=system_u:system_r:rhsmcertd_t:s0 tcontext=unconfined_u:object_r:gconf_home_t:s0 tclass=dir permissive=0
```

Note: I had to switch of `dontaudit` using `semodule --disable_dontaudit --build`, because there were lot of `dontaudit` for our daemon:

```
dontaudit rhsmcertd_t configfs_t:dir write;
dontaudit rhsmcertd_t defined_port_type:tcp_socket name_bind; [ nis_enabled ]:True
dontaudit rhsmcertd_t defined_port_type:udp_socket name_bind; [ nis_enabled ]:True
dontaudit rhsmcertd_t gnome_home_type:dir { getattr open search };
dontaudit rhsmcertd_t mountpoint:dir write;
dontaudit rhsmcertd_t nscd_t:fd use; [ nscd_use_shm ]:False
dontaudit rhsmcertd_t nscd_t:nscd { getserv shmemgrp shmemhost shmempwd shmemserv }; [ nscd_use_shm ]:False
dontaudit rhsmcertd_t nscd_var_run_t:file { getattr ioctl lock open read }; [ nscd_use_shm ]:False
dontaudit rhsmcertd_t nscd_var_run_t:file { getattr ioctl lock open read }; [ nscd_use_shm ]:True
dontaudit rhsmcertd_t port_type:tcp_socket name_bind; [ nis_enabled ]:True
dontaudit rhsmcertd_t port_type:udp_socket name_bind; [ nis_enabled ]:True
dontaudit rhsmcertd_t rhsmcertd_t:capability net_bind_service; [ nis_enabled ]:True
dontaudit rhsmcertd_t rhsmcertd_t:capability { dac_read_search net_admin sys_module };
dontaudit rhsmcertd_t rhsmcertd_t:file create;
dontaudit rhsmcertd_t rhsmcertd_t:udp_socket listen;
dontaudit rhsmcertd_t xguest_t:dbus send_msg;
```

Cheers,
Jiri

jirihnidek commented:
And here are AVC emssages, when system is not in Enforcing mode, but when it is in Permissive mode:

```
----
time->Thu Apr 11 10:55:24 2019
type=AVC msg=audit(1554972924.210:368): avc:  denied  { noatsecure } for  pid=2999 comm="virt-what" scontext=system_u:system_r:rhsmcertd_t:s0 tcontext=system_u:system_r:dmidecode_t:s0 tclass=process permissive=1
----
time->Thu Apr 11 10:55:24 2019
type=AVC msg=audit(1554972924.211:369): avc:  denied  { rlimitinh } for  pid=2999 comm="dmidecode" scontext=system_u:system_r:rhsmcertd_t:s0 tcontext=system_u:system_r:dmidecode_t:s0 tclass=process permissive=1
----
time->Thu Apr 11 10:55:24 2019
type=AVC msg=audit(1554972924.211:370): avc:  denied  { siginh } for  pid=2999 comm="dmidecode" scontext=system_u:system_r:rhsmcertd_t:s0 tcontext=system_u:system_r:dmidecode_t:s0 tclass=process permissive=1
----
time->Thu Apr 11 10:55:27 2019
type=AVC msg=audit(1554972927.223:371): avc:  denied  { noatsecure } for  pid=3069 comm="dbus-daemon-lau" scontext=system_u:system_r:system_dbusd_t:s0-s0:c0.c1023 tcontext=system_u:system_r:setroubleshootd_t:s0-s0:c0.c1023 tclass=process permissive=1
----
time->Thu Apr 11 10:55:27 2019
type=AVC msg=audit(1554972927.224:372): avc:  denied  { rlimitinh } for  pid=3069 comm="setroubleshootd" scontext=system_u:system_r:system_dbusd_t:s0-s0:c0.c1023 tcontext=system_u:system_r:setroubleshootd_t:s0-s0:c0.c1023 tclass=process permissive=1
----
time->Thu Apr 11 10:55:27 2019
type=AVC msg=audit(1554972927.224:373): avc:  denied  { siginh } for  pid=3069 comm="setroubleshootd" scontext=system_u:system_r:system_dbusd_t:s0-s0:c0.c1023 tcontext=system_u:system_r:setroubleshootd_t:s0-s0:c0.c1023 tclass=process permissive=1
```

WOnder93 commented:
@jirihnidek Hm, your patch adds permissions for `cert_t`, yet this type doesn't appear in either one of your AVCs...

jirihnidek commented:
@WOnder93 I know. Is it problem? Should I add permission for `container_config_t` instead? I don't see `container_config_t` in:

```
[root@fedora29 ~]# ls -lZ /etc/docker/certs.d/
total 0
drwxr-xr-x. 2 root root unconfined_u:object_r:cert_t:s0  69 Apr 11 10:55 access.redhat.com
drwxr-xr-x. 2 root root system_u:object_r:cert_t:s0     109 Apr 11 14:34 cdn.redhat.com
drwxr-xr-x. 2 root root system_u:object_r:cert_t:s0      27 Mar 28 14:12 redhat.com
drwxr-xr-x. 2 root root system_u:object_r:cert_t:s0      27 Mar 28 14:12 redhat.io
drwxr-xr-x. 2 root root unconfined_u:object_r:cert_t:s0  69 Apr 11 10:55 registry.access.redhat.com
drwxr-xr-x. 2 root root unconfined_u:object_r:cert_t:s0  69 Apr 11 10:55 registry.redhat.io
```

I can see there only `cert_t`.

jirihnidek commented:
@WOnder93 @wrabcak Hi, I updated the PR according your comments and I can confirm that it works now as expected.

wrabcak commented:
Hi Jiri, 
THanks for PR, however, fix should be somewhere else. I created new interface here:
https://github.com/containers/container-selinux/commit/ad5f6303c8d97abaae82159ba69bf2b937b10b40

So you need to add this lines:
```
optional_policy(`
    container_manage_config_files(rhsmcertd_t)
')
```

to rhsmcertd.te file.  :) 

Thanks,
Lukas

rhatdan commented:
container-selinux-2.97 will have the interface.

jirihnidek commented:
@wrabcak I updated the PR as requested, but I wasn't able to verify the bug fix, because "compiling" of `rhsmcertd.pp` was terminated with following error:

```
/usr/bin/checkmodule:  loading policy configuration from tmp/rhsmcertd.tmp
rhsmcertd.te:180:ERROR 'unknown type container_config_t' at token ';' on line 11321:
#line 180
	allow rhsmcertd_t container_config_t:dir { open read getattr lock search ioctl add_name remove_name write };
/usr/bin/checkmodule:  error(s) encountered while parsing configuration
```

despite I updated `/usr/share/selinux/devel/include/services/container.if` according your patch.

wrabcak commented:
I fixed container-selinux part of the fix and CI is passing. Merging. 

Thanks!
